### PR TITLE
[MOV] *_product_configurator,*: Merge * and *_product_configurator

### PR DESCRIPTION
**Pull request by chevalierv**
Since https://github.com/odoo/odoo/commit/ee3bf990d9160b085f5700a5ee69c781e9794fa2, `sale_product_configurator` is auto-installed. The
settings to enable or disable it have been removed and the only way to
deactivate the product configurator is to uninstall the module but then
it provides a poor UX (eg. in `website_sale` buy now button on /shop
page adds the first available combination to the cart instead of opening
the configurator).

Therefore, we have decided to merge modules to improve the readability
and reduce the complexity of our codebase.

See also:
- https://github.com/odoo/enterprise/pull/57577
- https://github.com/odoo/upgrade/pull/5750

task-3763473

### CODE REVIEW

File: addons/website_sale/data/demo.xml [hunk 1]
```diff
@@ -756,4 +756,13 @@
             <field name="user_id" ref="base.user_demo"/>
         </record>
 
+        <record id="sale.product_product_1_product_template" model="product.template">
+            <field name="website_sequence">9980</field>
+            <field name="is_published" eval="True"/>
+        </record>
+
+        <record id="product.product_product_4_product_template" model="product.template">
+            <field name="optional_product_ids" eval="[(6,0,[ref('product.product_product_11_product_template'), ref('website_sale.product_product_1_product_template')])]"/>
```
Comment (Feyensv) [inline addons/website_sale/data/demo.xml:765]: Command.set ? :D

File: addons/website_sale/tests/test_customize.py [hunk 2]
```diff
@@ -7,10 +7,10 @@
 from odoo.tools.misc import file_open
 
 from odoo.addons.base.tests.common import HttpCaseWithUserDemo, HttpCaseWithUserPortal
-
+from odoo.addons.sale.tests.product_configurator_common import TestProductConfiguratorCommon
 
 @tagged('post_install', '-at_install')
-class TestCustomize(HttpCaseWithUserDemo, HttpCaseWithUserPortal):
+class TestCustomize(HttpCaseWithUserDemo, HttpCaseWithUserPortal, TestProductConfiguratorCommon):
```
Comment (Feyensv) [inline addons/website_sale/tests/test_customize.py:13]: Mhhhhhhhm, the ProductConfiguratorCommon is doing way more than what you need.  Is it really needed to make that change ?

**Review by Feyensv — COMMENTED** [review]
didn't see anything strange :duck:

File: addons/website_sale/tests/test_customize.py (hunk 2 above)
Comment (chevalierv) [reply addons/website_sale/tests/test_customize.py:13]: The test I added in that file needs `TestProductConfiguratorCommon`, and the second test I changed in the same file was defining the same product that was already defined in `TestProductConfiguratorCommon`. I could just have created a new class and don't change anything here but I thought it was better to factorize that here :)

File: addons/website_sale/models/sale_order.py [hunk 3]
```diff
@@ -373,7 +373,14 @@ def _cart_update(self, product_id, line_id=None, add_qty=0, set_qty=0, **kwargs)
...
         }
 
-    def _cart_find_product_line(self, product_id, line_id=None, **kwargs):
+    def _cart_find_product_line(
+            self,
+            product_id,
+            line_id=None,
+            linked_line_id=False,
+            optional_product_ids=None,
+            **kwargs
```
Comment (Demesmaeker) [inline addons/website_sale/models/sale_order.py:382]:
```suggestion
            **kwargs,
```
Pure nitpicking, only if you have to change something else

File: addons/website_sale/models/sale_order.py [hunk 4]
```diff
@@ -397,7 +404,14 @@ def _cart_find_product_line(self, product_id, line_id=None, **kwargs):
         if line_id:
             domain += [('id', '=', line_id)]
         else:
-            domain += [('product_custom_attribute_value_ids', '=', False)]
+            domain += [
+                ('product_custom_attribute_value_ids', '=', False),
+                ('linked_line_id', '=', linked_line_id)
```
Comment (Demesmaeker) [inline addons/website_sale/models/sale_order.py:409]:
```suggestion
                ('linked_line_id', '=', linked_line_id),
```
same

**Comment by chevalierv** [comment]
> Hello!
> 
> Shouldn't all the tests regarding the product configurator now be in `sale_management` instead of `test_sale_product_configurators`? Also, I'm wondering if this logic should also be applied to the product matrix? Even if less used than the configurator, it's the other face of the same coin, IMHO.
> 
> Otherwise, lgtm, so I'll already green it =]
> 
> Have a nice day!

Hello :wave: 

Yes, we could also merge the test modules with their parent but I would rather take more time to access what test is testing what because I suspect that some tests are duplicates, and with the upcoming migration of the configurator to OWL in eCommerce, most of them will end up duplicated. I'll create a new task to merge those modules and refactor some of the codes.

The point of this PR was to move the code with the least changes possible to avoid behavior changes and to be fast. 

Have a nice day :)

**Comment by Feyensv** [comment]
> Shouldn't all the tests regarding the product configurator now be in `sale_management` instead of `test_sale_product_configurators`?

As they also consider the combination with event_sale and event_booth_sale.  I don't mind keeping a dedicated module ;).

**Comment by Feyensv** [comment]
@robodoo r+

**Pull request closed without merging**